### PR TITLE
NAV-220 Completed time for workflow state

### PR DESCRIPTION
**Pull request by tomeksabala**

### CODE REVIEW

File: clients/ckan_client.py [hunk 1]
```diff
@@ -113,7 +114,7 @@ class CkanError(Exception):
     app = create_app()
     with app.app_context():
         ckan_cli = init_ckan(apikey=os.getenv('CKAN_APIKEY'))
-        dataset_id = "3963fdf5-8915-448d-b8dd-9beca9c04a35"
-        fetch_workflow_state(ckan_cli, dataset_id)
+        dataset_id = "switzerland-country-estimates-2022"
+        state = fetch_workflow_state(ckan_cli, dataset_id)
```
Comment (jonathansberry) [inline clients/ckan_client.py:118]: Does this not break linting?  It looks like you create a variable and never use it.

File: logic.py [hunk 2]
```diff
@@ -109,3 +112,69 @@ def get_task_list_with_milestones(dataset_id, tasks, milestones):
...
+    new_state = copy.deepcopy(workflow_state)
+    completed_tasks = {task['id']: task for task in new_state["completedTasks"]}
+    if task_id not in completed_tasks:
+        raise LogicError(f"Task {task_id} not found in completed tasks")
+    new_state["completedTasks"].remove(completed_tasks[task_id])
+    return new_state
+
+
+def check_if_task_is_complete(task_id, wf_state):
+    completed_tasks = {task['id']: task for task in wf_state["completedTasks"]}
```
Comment (jonathansberry) [inline logic.py]: Why not just a plain set?

```suggestion
    completed_tasks = {task['id'] for task in wf_state["completedTasks"]}
```

File: logic.py [hunk 3]
```diff
@@ -109,3 +112,69 @@ def get_task_list_with_milestones(dataset_id, tasks, milestones):
...
+            default_created_time = 'Thu, 01 Jan 1970 00:00:00 +0000'
+            converted_completed_tasks.append({
+                'id': task_id,
+                'completedAt': default_created_time
+            })
+        else:
+            converted_completed_tasks.append(completed_task)
+    workflow_state["completedTasks"] = converted_completed_tasks
+    return workflow_state
+
```
Comment (jonathansberry) [inline logic.py]: Did we say this would be removed?

File: logic.py [hunk 4]
```diff
@@ -109,3 +112,69 @@ def get_task_list_with_milestones(dataset_id, tasks, milestones):
...
+
+def json_dumps(obj):
+    # uses flask builtin json serializer which converts :class:`datetime.datetime` and :class:`datetime.date` are
+    #   serialized to :rfc:`822` strings. This is the same as the HTTP date format.
+    # to decode you can use `email.utils.parsedate_to_datetime(date)`
```
Comment (jonathansberry) [inline logic.py]: Why not a multiline comment?
```suggestion
    """
    Uses flask builtin json serializer, for which :class:`datetime.datetime` and 
    :class:`datetime.date` are serialized to :rfc:`822` strings. This is the same 
    as the HTTP date format. 
    
    To decode you can use: `email.utils.parsedate_to_datetime(date)`
    """
```

File: tests/test_logic.py [hunk 5]
```diff
@@ -133,6 +134,69 @@ def test_is_task_completed_returns_false_for_unknown_task(logged_in):
...
+    actual_created = workflow_state['completedTasks'][0].get('completedAt')
+    assert now < actual_created
+
+
+def test_complete_task_is_idempotent(empty_workflow_state):
+    task_id = "Task1"
+    logic.complete_task(empty_workflow_state, task_id)
+    workflow_state = logic.complete_task(empty_workflow_state, task_id)
+    assert len(workflow_state['completedTasks']) == 1
+
```
Comment (jonathansberry) [inline tests/test_logic.py:156]: Ah interesting. Nice :+1:

**Review by jonathansberry — CHANGES_REQUESTED** [review]
Great job.  Just some questions/thoughts raised...

File: clients/ckan_client.py (hunk 1 above)
Comment (tomeksabala) [reply clients/ckan_client.py:118]: no, it's fine.

**Review by jonathansberry — APPROVED** [review]

**Pull request merged**